### PR TITLE
Merging mass-mapping to main

### PR DESCRIPTION
**Pull request by JessWhitney**
The full mass-mapping pipeline used for the paper, with updated notebooks and configs. No bugs as this worked from training all the way through to plotting/validation.

### CODE REVIEW

File: .DS_Store
Comment (tobias-liaudat) [inline .DS_Store:1]: This file needs to be removed

File: README.md [hunk 1]
```diff
@@ -1,72 +1,25 @@
...
-  year={2018}
-}
-
-@article{devries2019evaluation,
-  title={On the evaluation of conditional GANs},
-  author={DeVries, Terrance and Romero, Adriana and Pineda, Luis and Taylor, Graham W and Drozdzal, Michal},
-  journal={arXiv preprint arXiv:1907.08175},
-  year={2019}
-}
+This repository was forked from rcGAN by Bendel et al., with significant changes and modification made by Whitney et al.
```
Comment (tobias-liaudat) [inline README.md]: Add link to original Bendel paper as well as the original rcGAN repo

File: README.md [hunk 2]
```diff
@@ -1,72 +1,25 @@
-# A Regularized Conditional GAN for Posterior Sampling in Inverse Problems [[arXiv]](https://arxiv.org/abs/2210.13389)
+# Generative modelling for mass-mapping with fast uncertainty quantification [[arXiv]](https://arxiv.org/abs/2410.24197)
 ## Setup
```
Comment (tobias-liaudat) [inline README.md]: It'd be good to add a small description of the method (something like the abstract used for the paper but more synthetic). Also select one figure from the paper to illustrate the method (I'd suggest the one from the COSMOS field showing the reconstruction and std dev).

File: comments.md [hunk 3]
```diff
@@ -1,6 +1,14 @@
+# rcGAN development version  
+
 
 # Installation
 
+If in the Hypatia cluster, first run:
```
Comment (tobias-liaudat) [inline comments.md]: We should not mention Hypatia, but be general for a computer cluster.

File: README.md [hunk 4]
```diff
@@ -1,72 +1,25 @@
...
 See ```docs/setup.md``` for basic environment setup instructions.
```
Comment (tobias-liaudat) [inline README.md]: This file does not exist

File: README.md [hunk 5]
```diff
@@ -1,72 +1,25 @@
...
```
Comment (tobias-liaudat) [inline README.md:2]: You need to add a reference to the `comments.md` file where we explain how to install the model and some other useful stuff. There are two option, mention what you can find on that file and provide a link or merge the info of that file in the readme

File: docs/mass_mapping.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
+# Reproducing our MRI results
```
Comment (tobias-liaudat) [inline docs/mass_mapping.md]: Update MRI to mass mapping

File: README.md [hunk 7]
```diff
@@ -1,72 +1,25 @@
...
 
-## Reproducing our Results
-### MRI
-See ```docs/mri.md``` for instructions on how to setup and reproduce our MRI results.
-
-## Extending the Code
-See ```docs/new_applications.md``` for basic instructions on how to extend the code to your application.
+## Reproducing the our Results
```
Comment (tobias-liaudat) [inline README.md]: typo

File: README.md [hunk 8]
```diff
@@ -1,72 +1,25 @@
...
-}
-
-@misc{zeng2022github,
-    howpublished = {Downloaded from \url{https://github.com/zengxianyu/co-mod-gan-pytorch}},
-    month = sep,
-    author={Yu Zeng},
-    title = {co-mod-gan-pytorch},
-    year = 2022
-}
-```
```
Comment (tobias-liaudat) [inline README.md:37]: It'd be good to add a reproducibility section and links to Zenodo. It'd be good to find there the trained GAN used to generate the results. Also, some samples and reconstructions of the GAN. 

Should we include the simulations in a zenodo? It depends on the size of the sims. If they are too big, you could just upload some of them to have a representative set.

File: README.md [hunk 9]
```diff
@@ -1,72 +1,25 @@
...
 
-## Reproducing our Results
-### MRI
-See ```docs/mri.md``` for instructions on how to setup and reproduce our MRI results.
-
-## Extending the Code
-See ```docs/new_applications.md``` for basic instructions on how to extend the code to your application.
+## Reproducing the our Results
+### 
+See ```docs/mass_mapping.md``` for instructions on how to setup and reproduce our COSMOS results.
```
Comment (tobias-liaudat) [inline README.md:30]: Could you coordinate with Matthijs so that the radio application of the GAN is also updated on this main branch. It'd be good to have everything to reproduce the results of both papers in a single repo.

File: mass_map_utils/jobs/validate_test/plot_cGAN_mri.sh
Comment (tobias-liaudat) [inline mass_map_utils/jobs/validate_test/plot_cGAN_mri.sh:1]: Shouldn't the MRI job be on the `mri_utils` directory

File: mass_map_utils/jobs/training/training_cGAN_mri.sh
Comment (tobias-liaudat) [inline mass_map_utils/jobs/training/training_cGAN_mri.sh:1]: Shouldn't the MRI job be on the `mri_utils` directory?

File: mass_map_utils/jobs/validate_test/test_plot_cGAN_mri copy.sh
Comment (tobias-liaudat) [inline mass_map_utils/jobs/validate_test/test_plot_cGAN_mri copy.sh:1]: Why it finishes by ` copy.sh`?

Shouldn't the MRI job be on the `mri_utils` directory?

File: mass_map_utils/jobs/validate_test/val_plot_cGAN_mri.sh
Comment (tobias-liaudat) [inline mass_map_utils/jobs/validate_test/val_plot_cGAN_mri.sh:1]: Shouldn't the MRI job be on the `mri_utils` directory?

File: mass_map_utils/notebooks/GIF_maker.ipynb
Comment (tobias-liaudat) [inline mass_map_utils/notebooks/GIF_maker.ipynb:1]: It should be good to have a small readme file in the `mass_map_utils/notebooks/` directory such that there's a phrase describing each notebook in the directory.

File: utils/mri/transforms.py [hunk 10]
```diff
@@ -250,11 +250,49 @@ def normalize_instance(
     Returns:
         torch.Tensor: Normalized tensor
     """
+
     mean = data.mean()
     std = data.std()
 
     return normalize(data, mean, std, eps), mean, std
 
+def normalise_complex(
```
Comment (tobias-liaudat) [inline utils/mri/transforms.py:259]: You should add a docstring on this function and the one below to specify what those two default values represent. It would also be good to add a phrase describing what exactly each of the functions do.

File: train.py [hunk 11]
```diff
@@ -25,23 +27,33 @@ def load_object(dct):
...
+        cfg = json.loads(json.dumps(cfg), object_hook=load_object)
 
-        model = rcGAN(cfg, args.exp_name, args.num_gpus)
-    else:
-        print("No valid application selected. Please include one of the following args: --mri")
-        exit()
+        # Load the correct model
+        if cfg.experience == 'mri':
+            dm = MRIDataModule(cfg)
+            model = rcGAN(cfg, args.exp_name, args.num_gpus)
```
Comment (tobias-liaudat) [inline train.py:41]: It seems that the rcGAN disappeared from the `from models.lightning.rcGAN import rcGAN`

We should be consistent. Or we remove this from here or add the rcGAN module again so that it can be used as suggested by this line and the config files.

File: evaluation_scripts/cfid/cfid_metric.py [hunk 12]
```diff
@@ -152,7 +152,7 @@ def _get_embed_im(self, multi_coil_inp, mean, std, maps):
 
             S = sp.linop.Multiply((self.args.im_size, self.args.im_size), maps[i])
 
-            im = torch.tensor(S.H * tensor_to_complex_np(unnormal_im.cpu())).abs().cuda()
+            im = torch.real(torch.tensor(S.H * tensor_to_complex_np(unnormal_im.cpu()))).cuda()
```
Comment (tobias-liaudat) [inline evaluation_scripts/cfid/cfid_metric.py:157]: It'd be good to add a comment to the code to explain this modification. Eventually, leave the previous line commented to show the previous implementation.

Also, this function will probably be used in the radio problem, so it'd be good to check with Matthijs what makes more sense to use in his problem. This change could be a potential bug for his application if it goes unnoticed.

File: README.md (hunk 8 above)
Comment (JessWhitney) [reply README.md:37]: Say a user picks up our zenodo weights and wants to test everything works, should we create some basic script they can run given our model ckpt and a simulation?

File: README.md (hunk 9 above)
Comment (JessWhitney) [reply README.md:30]: Yes, will have to wait for Matthijs to pick this up I think, once he's got some time

File: train.py (hunk 11 above)
Comment (JessWhitney) [reply train.py:41]: I see the issue but I don't understand the suggestion. We do import rcGAN, and rcGAN is still in the modles/lightning folder. What exactly are you suggesting we change?

File: README.md (hunk 8 above)
Comment (tobias-liaudat) [reply README.md:37]: @JessWhitney Yes, definitely!

File: train.py (hunk 11 above)
Comment (tobias-liaudat) [reply train.py:41]: Ahh sorry, I think I missed reading a couple of lines when writing that comment. Feel free to ignore it

**Comment by JessWhitney** [comment]
@tobias-liaudat I believe all comments have now been addressed, this is ready for you to take a look at after the New Year :)

File: docs/mass_mapping.md [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+Install the required modules via anaconda:
+```
+conda create --name <env> --file requirements.txt
+```
+Note that you may need to change the PyTorch version, depending on your CUDA distribution.
+
+## Update Config
+Update ```configs/mass_map.yml``` with the path to your data, where you want to store checkpoints, and with the path to additional data such as masks.
+
+## Data Setup
```
Comment (tobias-liaudat) [inline docs/mass_mapping.md:15]: I think we need to add more detail to the data preparation description. It should be somehow clearer what steps/scripts/notebooks to use and in which order. It'd be good to link to the dataset sources we used, e.g. kappaTNG sims, and the COSMOS dataset.

**Review by tobias-liaudat — CHANGES_REQUESTED** [review]
Thanks @JessWhitney for the improvement. Everything is looking better now. I only have one comment regarding the description of the data preprocessing / input data handling that is not sufficiently described. 

By the way, great to see the Zenodo in place!

File: docs/mass_mapping.md (hunk 13 above)
Comment (JessWhitney) [reply docs/mass_mapping.md:15]: @tobias-liaudat ready for you to look at again, I've updated the data setup section

**Review by tobias-liaudat — APPROVED** [review]
Thanks @JessWhitney LGTM
